### ISSUE TITLE
Rewrite effect throttler

### ISSUE THREAD

**Issue by lorandszakacs** [busymachines/pure-movie-server#29]
Currently, `pms.effects.EffectThrottler` is a bit broken. It fails to release properly on failure, and it does not function properly upon cancellation. Who knows what other problems it has...

Needs a complete rewrite of the implementation, while maintaining the interface roughly the same. Although alternative strategies for setting an upper-bound on how many effects we can run in a given time interval are also accepted.

General specs:
- let `N` denote a number of requests
- let `I` denote a time interval
- `EffectThrottler.throttle[A](fa: F[A])` should ensure that a maximum of `N` effects of type F[A] are actually "run" within the given interval `I`. E.g. `N = 2` and `I = 1 second`, we call `throttle` 5 times, each `F[A]` takes, say, ~100ms each. Which means that only the first two effects are run, and the next two only after roughly `900ms`, and then another `900ms` pass until the 5th on is run.
- if any `fa` fails, then it should not affect the functioning of the `EffectThrottler` or any other effect that is being throttled.
- on cancellation, should ensure that whatever other effects are "waiting" should never run. And ensure that all resources are released properly.

Ideally the implementation should depend only on `cats-effect`. But depending on [`monix`](https://monix.io/) and/or [`fs2`](https://fs2.io/guide.html) is also acceptable if not possible otherwise.

[Some tests](https://github.com/busymachines/pure-movie-server/blob/master/pms-utils/effects/src/main/scala/pms/effects/EffectThrottler.scala) already exist... but better tests will have to be written for the `EffectThrottler`, including with effects that capture blocking IO (in tests it is enough to simulate with `Thread.sleep`). Use [`TestContext`](https://github.com/typelevel/cats-effect/blob/1846813109b1e78c5bf36e6e179d7a91419e01d0/laws/shared/src/main/scala/cats/effect/laws/util/TestContext.scala) from the `cats-laws` library (we do not depend on it yet, but it's fine to add) to simulate passage of time when needed.

Recommended reading:
[cats-effect](https://typelevel.org/cats-effect/datatypes/)
[cats-effect concurrency](https://typelevel.org/cats-effect/concurrency/) — current implementation uses a `Semaphore`, but that can be changed as needed.
[cats-effect timer](https://typelevel.org/cats-effect/datatypes/timer.html)

Talks about concurrency in cats-effects:
- [Scale By The Bay 2018: Michael Pilquist, Concurrency with Cats-effect
(https://www.youtube.com/watch?v=Gig-f_HXvLI)
- [Composable concurrency with Ref + Deferred – Fabio Labella
](https://www.youtube.com/watch?v=-dLp3u6y2DQ)

**Comment by alexandrustana** [comment]
I think this can be closed :)
